drivers/rtt_rtc/rtt_rtc.c

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (C) 2020 Benjamin Valentin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     drivers_periph_rtc
+ * @{
+ *
+ * @file
+ * @brief       Basic RTC implementation based on a RTT
+ *
+ * @note        Unlike a real RTC, this emulated version is not guaranteed to keep
+ *              time across reboots or deep sleep.
+ *              If your hardware provides the means to implement a real RTC, always
+ *              prefer that over this emulated version!
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ * @}
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "periph/rtc.h"
+#include "periph/rtt.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define RTT_SECOND      (RTT_FREQUENCY)
+
+#define _RTT(n)         ((n) & RTT_MAX_VALUE)
+
+#define RTT_SECOND_MAX  (RTT_MAX_VALUE/RTT_FREQUENCY)
+
+#define TICKS(x)        (    (x) * RTT_SECOND)
+#define SECONDS(x)      (_RTT(x) / RTT_SECOND)
+
+#define _NOINIT         __attribute__((section(".noinit")))
+
+static uint32_t alarm_time;     /*< The RTC timestamp of the (user) RTC alarm */
+static uint32_t rtc_now _NOINIT;/*< The RTC timestamp when the last RTT alarm triggered */
+
+static uint32_t last_alarm;     /*< The RTT timestamp of the last alarm */
+
+static rtc_alarm_cb_t alarm_cb; /*< RTC alarm callback */
+static void *alarm_cb_arg;      /*< RTC alarm callback argument */
+
+static void _rtt_alarm(void *arg);
+
+/* convert RTT counter into RTC timestamp */
+static inline uint32_t _rtc_now(uint32_t now)
+{
+    return rtc_now + SECONDS(now - last_alarm);
+}
+
+static inline void _set_alarm(uint32_t now, uint32_t next_alarm)
+{
+    rtt_set_alarm(now + next_alarm, _rtt_alarm, NULL);
+}
+
+/* This calculates when the next alarm should happen.
+   Always chooses the longest possible period min(alarm, overflow)
+   to minimize the ammount of wake-ups. */
+static void _update_alarm(uint32_t now)
+{
+    uint32_t next_alarm;
+
+    last_alarm = TICKS(SECONDS(now));
+
+    /* no alarm or alarm beyond this period */
+    if ((alarm_cb == NULL)     ||
+        (alarm_time < rtc_now) ||
+        (alarm_time - rtc_now > RTT_SECOND_MAX)) {
+        next_alarm = RTT_SECOND_MAX;
+    } else {
+        /* alarm triggers in this period */
+        next_alarm = alarm_time - rtc_now;
+    }
+
+    /* alarm triggers NOW */
+    if (next_alarm == 0) {
+        next_alarm = RTT_SECOND_MAX;
+        alarm_cb(alarm_cb_arg);
+    }
+
+    _set_alarm(now, TICKS(next_alarm));
+}
+
+/* the RTT alarm callback */
+static void _rtt_alarm(void *arg)
+{
+    (void) arg;
+
+    uint32_t now = rtt_get_counter();
+    rtc_now = _rtc_now(now);
+
+    _update_alarm(now);
+}
+
+void rtc_init(void)
+{
+    last_alarm = rtt_get_counter();
+    _set_alarm(last_alarm, TICKS(RTT_SECOND_MAX));
```

Comment:
With this implementation I encounter the following problem with ESPs. `last_alarm` is set here to the current RTT counter value. The calculation of the real time in seconds is done with `rtc_now + SECONDS(now - last_alarm)`. This means that the current counter value corresponds to `rtc_now` here. Therefore, the real time always restarts at the value as last set by the RTT alarm handler or with `rtc_set_time` after a reboot or deep sleep, i.e. again and again from `rtc_now`.

Translated with www.DeepL.com/Translator (free version)
```
> rtc gettime
2020-01-01 00:01:06
> rtc reboot
> rtc gettime
2020-01-01 00:00:05
...
> rtc settime 2020-03-16 11:42:45
> rtc gettime
2020-03-16 11:45:20
> reboot
> rtc gettime
> 2020-03-16 11:42:48
> rtc setalarm 2020-03-16 11:44:00
> The alarm rang
> reboot
> rtc gettime
> 2020-03-16 11:44:00
```
I could solve the problem for ESPs as following:
Place `rtc_now` AND `load_alarm` in a memory section that is initialized to 0 at power-up but is retained during reboot or deep sleep. For ESPs this is `.rtc.bss`. Change the implementation as following
```suggestion
    if (last_alarm == 0) {
        uint32_t now = rtt_get_counter();
        rtc_now = _rtc_now(now);
        last_alarm = now;
    }
```
That is, these variables are only set during start-up after power on.

Reply:
Why is the `rtc_now = _rtc_now(now);` needed? `rtc_now` starting from 0 should be fine in the cold-boot case.

Reply:
Yes, setting `last_alarm` is of course sufficient.